acme/db/nosql/account.go

```diff
@@ -177,55 +178,106 @@ func (db *DB) CreateExternalAccountKey(ctx context.Context, name string) (*acme.
+		ProvisionerName: provisionerName,
+		Name:            name,
+		KeyBytes:        random,
+		CreatedAt:       clock.Now(),
 	}
 
 	if err = db.save(ctx, keyID, dbeak, nil, "external_account_key", externalAccountKeyTable); err != nil {
 		return nil, err
 	}
 	return &acme.ExternalAccountKey{
-		ID:        dbeak.ID,
-		Name:      dbeak.Name,
-		AccountID: dbeak.AccountID,
-		KeyBytes:  dbeak.KeyBytes,
-		CreatedAt: dbeak.CreatedAt,
-		BoundAt:   dbeak.BoundAt,
+		ID:              dbeak.ID,
+		ProvisionerName: dbeak.ProvisionerName,
+		Name:            dbeak.Name,
+		AccountID:       dbeak.AccountID,
+		KeyBytes:        dbeak.KeyBytes,
+		CreatedAt:       dbeak.CreatedAt,
+		BoundAt:         dbeak.BoundAt,
 	}, nil
 }
 
 // GetExternalAccountKey retrieves an External Account Binding key by KeyID
-func (db *DB) GetExternalAccountKey(ctx context.Context, keyID string) (*acme.ExternalAccountKey, error) {
+func (db *DB) GetExternalAccountKey(ctx context.Context, provisionerName string, keyID string) (*acme.ExternalAccountKey, error) {
 	dbeak, err := db.getDBExternalAccountKey(ctx, keyID)
 	if err != nil {
 		return nil, err
 	}
 
+	if dbeak.ProvisionerName != provisionerName {
+		return nil, acme.NewError(acme.ErrorUnauthorizedType, "name of provisioner does not match provisioner for which the EAB key was created")
+	}
+
 	return &acme.ExternalAccountKey{
-		ID:        dbeak.ID,
-		Name:      dbeak.Name,
-		AccountID: dbeak.AccountID,
-		KeyBytes:  dbeak.KeyBytes,
-		CreatedAt: dbeak.CreatedAt,
-		BoundAt:   dbeak.BoundAt,
+		ID:              dbeak.ID,
+		ProvisionerName: dbeak.ProvisionerName,
+		Name:            dbeak.Name,
+		AccountID:       dbeak.AccountID,
+		KeyBytes:        dbeak.KeyBytes,
+		CreatedAt:       dbeak.CreatedAt,
+		BoundAt:         dbeak.BoundAt,
 	}, nil
 }
 
-func (db *DB) UpdateExternalAccountKey(ctx context.Context, eak *acme.ExternalAccountKey) error {
+func (db *DB) DeleteExternalAccountKey(ctx context.Context, keyID string) error {
+	_, err := db.db.Get(externalAccountKeyTable, []byte(keyID))
+	if err != nil {
+		return errors.Wrapf(err, "error loading ACME EAB Key with Key ID %s", keyID)
+	}
+	err = db.db.Del(externalAccountKeyTable, []byte(keyID))
+	if err != nil {
+		return errors.Wrapf(err, "error deleting ACME EAB Key with Key ID %s", keyID)
+	}
+	return nil
+}
+
+// GetExternalAccountKeys retrieves all External Account Binding keys for a provisioner
+func (db *DB) GetExternalAccountKeys(ctx context.Context, provisionerName string) ([]*acme.ExternalAccountKey, error) {
+	entries, err := db.db.List(externalAccountKeyTable)
+	if err != nil {
+		return nil, err
+	}
+
+	keys := make([]*acme.ExternalAccountKey, len(entries))
+	for i, entry := range entries {
+		dbeak := new(dbExternalAccountKey)
+		if err = json.Unmarshal(entry.Value, dbeak); err != nil {
+			return nil, errors.Wrapf(err, "error unmarshaling external account key %s into dbExternalAccountKey", string(entry.Key))
+		}
+		keys[i] = &acme.ExternalAccountKey{
```

Comment:
Shouldn't we be filtering against the input provisionerName?

Reply:
Fixed in 02cd3b6b3bf763662b24bef35279b07dd000b107.